File: xls-sql-converter/src/main/java/ru/yandex/market/v1/Converter.java
package ru.yandex.market.v1;

import java.util.List;

import ru.yandex.market.v1.reader.XlsReader;
import ru.yandex.market.v1.table.Table;
import ru.yandex.market.v1.writer.DbDataWriter;
import ru.yandex.market.v1.writer.DbUnitXmlWriter;
import ru.yandex.market.v1.writer.SqlWriter;

public class Converter {
    private static final String DEFAULT_TARGET_SQL = "output.sql";
    private static final String DEFAULT_TARGET_DB_UNIT_XML = "output.xml";
    private final String source;
    private final List<Table> tables;

    private Converter(String source, List<Table> tables) {
        this.source = source;
        this.tables = tables;
    }

    public static Converter read(String inputPath) {
        return new Converter(inputPath, new XlsReader(inputPath).read());
    }

    public Converter saveAsSql(String target) {
        return save(new SqlWriter(), target);
    }

    public Converter saveAsSql() {
        return saveAsSql(DEFAULT_TARGET_SQL);
    }

    public Converter saveAsDbUnitXml(String target) {
        return save(new DbUnitXmlWriter(), target);
    }

    public Converter saveAsDbUnitXml() {
        return saveAsDbUnitXml(DEFAULT_TARGET_DB_UNIT_XML);
    }

    private Converter save(DbDataWriter writer, String target) {
        if (source.equalsIgnoreCase(target)) {
            throw new IllegalArgumentException("Target file is source file");
        }
        writer.write(tables, target);
        return this;
    }
}

File: xls-sql-converter2/src/main/kotlin/app/converter/Converter.kt
package app.converter

import app.converter.reader.XlsReader
import app.converter.table.Table
import app.scriptexecutor.ScriptExecutor
import app.converter.writer.DbDataWriter
import app.converter.writer.SqlWriter
import org.springframework.beans.factory.config.ConfigurableBeanFactory.SCOPE_PROTOTYPE
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component

@Scope(SCOPE_PROTOTYPE)
@Component
class Converter(private val scriptExecutor: ScriptExecutor) {
    companion object {
        private const val DEFAULT_TARGET_SQL = "output.sql"
    }

    private val sources = mutableListOf<String>()
    private val tables = mutableListOf<Table>()
    private var sqlScriptPath: String? = null

    fun read(inputPath: String): Converter {
        sources.add(inputPath)
        tables.addAll(XlsReader(inputPath).read())
        sqlScriptPath = null
        return this
    }

    fun saveAsSql(target: String = DEFAULT_TARGET_SQL): Converter {
        save(SqlWriter(), target)
        sqlScriptPath = target
        return this
    }

    fun applyToDatabase(): Converter {
        if (sqlScriptPath === null) {
            saveAsSql()
        }
        scriptExecutor.execute(sqlScriptPath!!)
        return this
    }

    private fun save(writer: DbDataWriter, target: String): Converter {
        require(sources.none { it.equals(target, ignoreCase = true) }) { "Target file is source file" }
        writer.write(tables, target)
        return this
    }
}